eth2near/contract_wrapper/src/dao_contract.rs

```diff
@@ -0,0 +1,216 @@
+use near_sdk::{AccountId, Balance, Gas};
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use std::collections::HashMap;
+use std::error::Error;
+
+use crate::contract_wrapper_trait::ContractWrapper;
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
+#[serde(crate = "near_sdk::serde")]
+pub struct ProposalOutput {
+    /// Id of the proposal.
+    pub id: u64,
+    #[serde(flatten)]
+    pub proposal: Proposal,
+}
+
+/// Status of a proposal.
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Clone, PartialEq, Debug)]
+#[serde(crate = "near_sdk::serde")]
+pub enum ProposalStatus {
+    InProgress,
+    /// If quorum voted yes, this proposal is successfully approved.
+    Approved,
+    /// If quorum voted no, this proposal is rejected. Bond is returned.
+    Rejected,
+    /// If quorum voted to remove (e.g. spam), this proposal is rejected and bond is not returned.
+    /// Interfaces shouldn't show removed proposals.
+    Removed,
+    /// Expired after period of time.
+    Expired,
+    /// If proposal was moved to Hub or somewhere else.
+    Moved,
+    /// If proposal has failed when finalizing. Allowed to re-finalize again to either expire or approved.
+    Failed,
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
+#[cfg_attr(not(target_arch = "wasm32"), derive(Clone, Debug))]
+#[serde(crate = "near_sdk::serde")]
+pub struct ActionCall {
+    pub method_name: String,
+    pub args: Base64VecU8,
+    pub deposit: U128,
+    pub gas: U64,
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
+#[cfg_attr(not(target_arch = "wasm32"), derive(Clone, Debug))]
+#[serde(crate = "near_sdk::serde")]
+pub enum ProposalKind {
+    /// Add member to given role in the policy. This is short cut to updating the whole policy.
+    AddMemberToRole { member_id: AccountId, role: String },
+    /// Remove member to given role in the policy. This is short cut to updating the whole policy.
+    RemoveMemberFromRole { member_id: AccountId, role: String },
+    /// Calls `receiver_id` with list of method names in a single promise.
+    /// Allows this contract to execute any arbitrary set of actions in other contracts.
+    FunctionCall {
+        receiver_id: AccountId,
+        actions: Vec<ActionCall>,
+    },
+    /// Indicates that given bounty is done by given user.
+    BountyDone {
+        bounty_id: u64,
+        receiver_id: AccountId,
+    },
+    /// Just a signaling vote, with no execution.
+    Vote,
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Clone, Debug)]
+#[serde(crate = "near_sdk::serde")]
+pub enum Vote {
+    Approve = 0x0,
+    Reject = 0x1,
+    Remove = 0x2,
+}
+
+/// Proposal that are sent to this DAO.
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
+#[cfg_attr(not(target_arch = "wasm32"), derive(Debug))]
+#[serde(crate = "near_sdk::serde")]
+pub struct Proposal {
```

Comment:
I think It is not exactly `Proposal`. `ProposalInput` looks more like Proposal. It is more looks like Proposal with a lot of supporting information. May be `ProposalState` is better?

Reply:
All types are migrated from here, so let's keep the same naming 
https://github.com/near-daos/sputnik-dao-contract/blob/main/sputnikdao2/src/proposals.rs